doc/internal/consensus/Consensus-Protocol.md

```diff
@@ -0,0 +1,127 @@
+# The consensus
+
+Before talking about consensus, we need to discuss some logistics based on how the systems can co-exist.
+
+* Communication: Distributed systems need a method to communicate between each other. Remote Procedure Calls is the mechanism using which a standalone server can talk to another. The existing `network` layer of the database will handle all the communication between servers. 
+* Security: Access control mechanisms need to be in place to decide on access to functions in the servers based on their state (leader, follower, candidate)
+* Routing to leader: One of the issues with a varying leader is for the clients to know which IP address to contact for the service. We can solve this problem by advertising any/all IPs of the cluster and simply forward this request to the current leader; OR have a proxy that can forward the request to the current leader wheneve the requests come in. (Section client interaction of post has another approach which works too)
```

Comment:
As this is a design doc, please make a decision, and state, why you made that decision, and why you didn't choose the other solution.

Personally, I probably would make the discovery of all other nodes part of the cluster login handshake, but it's your decision.

Reply:
The cluster login handshake is an initial setup problem, yes.

But here since the leaders keep changing, we need another method to figure out who the leader is.

I'll stick to the Client providing the IP one, seems better.